Episode 1: derived data types (#40)
ENCCS/intermediate-mpi: content/derived-datatypes.rst

.. objectives::

   - Understand how MPI handles datatypes.

>    - Learn to send and receive composite messages

Reviewer: ```suggestion
   - Learn to send and receive messages using composite datatypes
```

>    - Learn how to represent homogeneous collections as MPI messages. |term-MPI_Type_contiguous|, |term-MPI_Type_vector|, |term-MPI_Type_indexed|

Reviewer: ```suggestion
   - Learn how to represent homogeneous collections as MPI datatypes. |term-MPI_Type_contiguous|, |term-MPI_Type_vector|, |term-MPI_Type_indexed|
```

>    - Learn how to represent your own derived datatypes as MPI messages with |term-MPI_Type_create_struct| and |term-MPI_Type_commit|

Reviewer: ```suggestion
   - Learn how to represent your own derived datatypes as MPI datatypes with |term-MPI_Type_create_struct| and |term-MPI_Type_commit|
```

The ability to define custom datatypes is one of the hallmarks of a modern
programming language, since it allows programmers to structure their code in a
way that enhances readability and maintainability.
How can this be done in MPI? Recall that MPI is a standard describing a library
to enable parallel programming in the message passing model.

> MPI supports many of the basic datatypes recognized by the C standard.

Reviewer: ```suggestion
MPI supports many of the basic datatypes recognized by the C and Fortran standards.
```

   ``MPI_WCHAR`` ; ``wchar_t``
   ``MPI_SHORT`` ; ``short``
   ``MPI_INT`` ; ``int``
   ``MPI_LONG`` ; ``long``
   ``MPI_LONG_LONG_INT`` ; ``long long``
   ``MPI_SIGNED_CHAR`` ; ``signed char``
   ``MPI_UNSIGNED_CHAR`` ; ``unsigned char``
   ``MPI_UNSIGNED_SHORT`` ; ``unsigned short``
   ``MPI_UNSIGNED`` ; ``unsigned int``
   ``MPI_UNSIGNED_LONG`` ; ``unsigned long``
   ``MPI_UNSIGNED_LONG_LONG`` ; ``unsigned long long``
   ``MPI_C_COMPLEX`` ; ``float _Complex``
   ``MPI_C_DOUBLE_COMPLEX`` ; ``double _Complex``
   ``MPI_C_LONG_DOUBLE_COMPLEX`` ; ``long double _Complex``
   ``MPI_PACKED`` ;
   ``MPI_BYTE`` ;

In the C language, types are **primitive** constructs: they
are *defined* by the standard and *enforced* by the compiler.
The MPI types are instead **variants** in the ``MPI_Datatype`` enumeration: they
appear as the **same** type to the compiler.
This is a fundamental difference which influences the way custom datatypes are handled.

In the C language, you would declare a ``struct`` such as the following:

.. code-block:: c

   struct Pair {
    int first;
    char second;
   };

``Pair`` is a new type. From the compiler's point of view, it has status on par

> with the fundamental datatypes introduced above. The C standard mandates *how

Reviewer: ```suggestion
with the fundamental datatypes introduced above. The C standard makes requirements on *how
```

since details like alignment requirements are implementation-specific. Memory layout of composite types in C is not universally portable. (If it was, then MPI would not need to jump through hoops at run time to learn about the layout!)

Author: Indeed! :smile_cat:

     \textrm{LB}[m] = \min_{j}[\textrm{Displacement}_{j}]

Upper bound
  The last byte occupied by the datatype.

  .. math::
     :label: eq:upperbound

     \textrm{UB}[m] = \max_{j}[\textrm{Displacement}_{j} + \texttt{sizeof}(\textrm{Datatype}_{j})] + \textrm{Padding}

Extent
  The amount of memory needed to represent the datatype, taking into account architecture-specific alignment.

  .. math::
     :label: eq:extent

     \textrm{Extent}[m] = \textrm{UB}[m] - \textrm{LB}[m]

The C language (and Fortran) *require* that the data occurs in memory at
well-defined addresses: the data needs to be aligned. The address, in bytes, of
any item must be a multiple of the size of that item in bytes. This is so-called
*natural alignment*.
For our ``Pair`` data structure the ``first`` element is an ``int`` and occupies
4 bytes. An ``int`` will align to 4 bytes boundaries: when allocating a new
``int`` in memory, the compiler will insert **padding** to reach the alignment
boundary.
Indeed, ``second`` is a ``char`` and requires just 1 byte. To insert yet another
``Pair`` item, we first need to reach the alignment boundary with a padding of 3
bytes.
Thus:

.. math::

>    \begin{aligned}

Reviewer: To be even more clear, you could enumerate the values of Displacement_j and sizeof(Datatype_j), so that the different values of 4 on line 178 are clearly understandable

Author: Good idea.

---------------------

MPI offers the possibility to pack and unpack data of known datatype into a
single contiguous memory buffer, *without* first having to define a
corresponding datatype.
This can be an extremely useful technique to reduce messaging traffic and could
help with the readability and portability of the code.
The resulting packed buffer will be of type ``MPI_PACKED`` and can contain any
sort of heterogeneous collection of basic datatypes recognized by MPI.

.. figure:: img/E01-pack_unpack.svg
   :align: center

   MPI allows the programmer to communicate heterogeneous collections into a
   single message, without defining a full-fledged custom datatype. The data is
   packed into a buffer of type ``MPI_PACKED``. On the receiving end, the buffer
   will be unpacked into its constituent components.

.. signature:: |term-MPI_Pack|

   Pack data in noncontiguous memory to a contiguous memory buffer.

   .. code-block:: c

      int MPI_Pack(const void *inbuf,
                   int incount,
                   MPI_Datatype datatype,
                   void *outbuf,
                   int outsize,
                   int *position,
                   MPI_Comm comm)

> .. figure:: img/E01-pack.svg

Reviewer: I don't understand what the pink color represents (or indeed the whoel graphic). Perhaps separating input and output depictions would help?

Author: Pink is supposed to be data already packed in `outbuf`. Looking at the picture again, I agree it's confusing. I can reword the caption.

Author: Figures remade and with a (hopefully) better worded caption!

   :align: center

   Schematic view of the relation between ``inbuf``, ``outbuf``, and the
   ``position`` input/output parameter.

.. parameters::

   ``inbuf``
     The input buffer, *i.e.* the data to be packed into contigous memory.
   ``incount``
     Number of input data items.
   ``datatype``
     The datatype of each item to be packed.
   ``outbuf``
     Starting address of the output buffer.
   ``outsize``
     The size, in bytes, of the output buffer.
   ``position``
     This is an input/output parameter:

     - In *input*, the data in ``inbuf`` will be copied at the address

>        ``outbuf + *position``.

Reviewer: Did you mean inbuf, rather than outbuf?

Author: Double-checked and it's correct as written here. Do you think the following phrasing might be clearer:
```rst
This is an input/output parameter to keep track of where the current and next, if any, elements should be packed in ``outbuf``:

- In *input*, the data in ``inbuf`` will be copied in ``outbuf`` starting at address ``outbuf + *position``.
- In *output*, it is the first location in ``outbuf`` *after* copying the data: the next element, if any, to be packed will start here. 
```

Author: But it is definitely wrong for ``MPI_Unpack``!

Reviewer: How about

```
   ``position``
     This is an input/output parameter that describes locations within ``outbuf``. The data at ``inbuf`` will be copied to
     ``outbuf + *position``. After the function returns, the value ``*position`` indicates the first position in ``outbuf``
      that follows the data just copied. (This is useful to pass as ``position`` to the next call to ``MPI_Pack``.)
```

Author: :+1: I am redrawing the figure as before-and-after. Together with a better caption it should drive the point home.

.. figure:: img/E01-pack.svg
   :align: center

   The relation of ``inbuf``, ``outbuf``, and ``position`` when calling
   ``MPI_Pack``.  In this figure, ``outbuf`` already holds some data (the red
   shaded area). The data in ``inbuf`` is copied to ``outbuf`` starting at the

>    address given with ``position``.  When the function returns, the ``position``

Reviewer: Technically position is an offset into the buffer. An address is the value of a pointer.

```suggestion
   address ``outbuf+*position``.  When the function returns, the ``position``
```

>    parameter will have been updated to the *first* position in ``outbuf``

Reviewer: ```suggestion
   parameter will have been updated to refer to the *first* position in ``outbuf``
```

>    following the just copied data.

Reviewer: ```suggestion
   following the data copied by this call.```

- General types: |term-MPI_Type_create_struct|
- Register and unregister your types: |term-MPI_Type_commit|, |term-MPI_Type_free|

.. figure:: img/E01-type-life-cycle.svg
   :align: center

   The lifecycle of user-defined datatypes in MPI. Calling any of the type
   constructors will create an object of type ``MPI_Datatype`` with the
   user-defined typemap. Before using this custom datatype in message passing,
   it needs to be published with |term-MPI_Type_commit|: the typemap is made
   known to the runtime, allowing it to handle messages of the new custom type.

>    The programmer must take care to free the custom datatype object.

Reviewer: Suggest also spelling out the advantages and disadvantages vs packing
